lib/Partitioner/Partitioner.cpp

```diff
@@ -0,0 +1,365 @@
+      Node *N = bfs.levels[current].second[i];
+
+      for (int j = 0, e = N->getNumInputs(); j < e; ++j) {
+        Node *in = N->getNthInput(j).getNode();
+        if (isa<Storage>(in) || bfs.visited.count(in)) {
+          continue;
+        }
+        nList.push_back(in);
+        bfs.visited.insert(in);
+      }
+    }
+    if (nList.size() > 0) {
+      auto newPair = std::make_pair(level, nList);
+      bfs.levels.push_back(newPair);
+      level++;
+    }
+    current++;
+  }
+
+  return bfs;
+}
+
+/// Assign nodes to partitions and return the mapping.
+NodeFunctionMap Partitioner::selectPartitions(Function *F,
+                                              unsigned availableMemory) {
+  NodeFunctionMap mapping;
+  BFSLevel bfs = getBFSLevel(F);
+  unsigned level = bfs.levels.size();
+  std::vector<int>
+      cut; // A list of cut. The graph can be partitioned by levels [level - 1,
+           // cut[0]), [cut[0] - 1, cut[1]), ..., [cut[n], -1).
+
+  // Step 1 : get the initial cut based on BFS levels and avaiableMemory.
+  // TODO .. need to remove the duplicated memory useage.
+  unsigned mem = 0;
+  for (int i = level - 1; i >= 0; i--) {
+    unsigned tmp = 0;
+    for (int j = 0, e = bfs.levels[i].second.size(); j < e; j++) {
+      Node *N = bfs.levels[i].second[j];
+      tmp += memUsage[N];
+    }
+    if (mem + tmp > availableMemory) {
+      if (mem == 0) {
+        // This means the mem usage for one level exceeds the availableMem,
+        // accept it now and will do adjustment later.
+        cut.push_back(i + 1);
+      } else {
+        cut.push_back(i);
+        mem = tmp;
+      }
+    } else {
+      mem += tmp;
+    }
+  }
+
+  // The last border.
+  cut.push_back(-1);
+
+  // Step 2 : Create the initial mapping between node and functions.
+  for (int k = 0, e = cut.size(); k < e; k++) {
+    auto *newF = F->getParent()->createFunction(std::string(F->getName()) +
+                                                "_part" + std::to_string(k));
+    mapping.create(newF);
+    for (int i = k > 0 ? cut[k - 1] : level - 1; i > cut[k]; i--) {
+      for (int j = 0, e1 = bfs.levels[i].second.size(); j < e1; j++) {
+        Node *N = bfs.levels[i].second[j];
+        mapping.add(N, newF);
+      }
+    }
+  }
+  // Step 3 : adjust the partition based on performance (Advanced Graph
+  // Paritioning algrithm will be applied here).
+  // --- TODO
+
+  return mapping;
+}
+
+/// Adjust the logicalDevice ID for each DAGNode. This happens when \p num (i.e.
+/// the number of DAGNodes) is larger than the number of devices. E.g:
+/// node1(6GB) -> node2(14GB) -> node3(6GB). The memory limitation is 16GB, and
+/// there is only 2 devices.
+void Partitioner::adjustLogicalDevicdID(DAGNode *DAG, int num) {}
+
+/// Current only partition the representive function.
+DAGNode *Partitioner::doPartitioning(Function *F, NodeFunctionMap &mapping) {
+  // The dummy node.
+  DAGNode *DAG = new DAGNode();
```

Comment:
can you return std::unique ptr to avoid any memory leak issues?

these days we should avoid naked `new` with all the support we have in std:: various pointers.

Reply:
@rdzhabarov After discussing with @gcatron, we use shared_ptr since all DAGNodes are in HM.

Reply:
Sounds good.

Reply:
Generally speaking, and I am not sure if this applies to this case, I don't like shared_ptr because this is usually an indication that there is no clear ownership of some data structures because the lifetime of the objects are too complicated to analyze. Let's make sure that the design is simple enough for us to use regular pointers and containers with clear and simple lifetime.

Reply:
@nadavrot we don't use shared_ptr now.  This is the current design: https://github.com/pytorch/glow/pull/2268#issuecomment-455456113